### ISSUE TITLE
Too much CPU and RAM usage

### ISSUE THREAD

**Issue by NN---** [gitextensions/gitextensions#6732]
## Current behaviour
GitExtensions comsumes 4GB RAM, 100% CPU and spawns a lot of git.exe proceses, some of them don't quit.
With 3 GitExtensions windows opened and 3 VS I had 32 GB (100%) RAM usage..
I could kill all ~1000 living processes, there was one with PID of 190000 !

## Expected behaviour
Not consume so much resources.

## Steps to reproduce
I have a project of 1000 C++ files, compiling it while GitExtensions is opened seems like causing a problem.

## Did this work in previous version of GitExtensions
Don't know

## Environment
- GitExtensions version: 3.1
- GIT version: 2.21
- OS version: Windows 10
- .NET version:  4.8

## Diagnostics
Next time it happens I'll try making a dump but the machine was so slow this time that I barely could run cmd

**Comment by mstv** [comment]
I've seen this, too, once in the past two weeks. GE was spawning Git.exe instances one after the other without noticeable delay.
After I had killed the whole process tree of GE, it did not happen again although I built systems of similar size many times.

Also see the discussion in #6521.

**Comment by gerhardol** [comment]
To investigate at least repo structure, size, commit is needed and especially the Git Command Log (F12).
See the wiki https://github.com/gitextensions/gitextensions/wiki/Diagnosing-problems

Note that what is discussed in https://github.com/gitextensions/gitextensions/issues/6521#issuecomment-489331178 is not applicable here, that is only one Git process for one third of the time.

**Comment by mstv** [comment]
Is there an option to write the Git Command Log to a file all the time? It's very hard to do this manually when GE loads the PC to 100%.

Unfortunately, I cannot see what you have described. Usually, no Git.exe is spawned during builds. And if it was spawned, there were no delays inbetween,

Repo structure:
.gitignore = /bin/
bin/
src/tree...

CMake/Ninja generate and compile 2000+ C and C++ files.

**Comment by gerhardol** [comment]
> Is there an option to write the Git Command Log to a file all the time? 

Added to #4753

**Comment by NN---** [comment]
It may be related to TortoiseGit running.
I'll try removing all components to get a better reproduction.

**Comment by NN---** [comment]
I have opened a folder and have a lot of git.exe processes and seems like they never stops spawning.
Had to close forcibly all git.exe manually
I have a dump of GitExtensions if this helps.

**Comment by RussKie** [comment]
Is this issue affecting only this repo?
Is this happening only on this machine?

> I have a dump of GitExtensions if this helps.

It's not something I've looked at for a long time, but sure. I hope there is nothing super-sensitive in it.
If you can upload it somewhere and email the link to gitextensions -at- outlook dot com, we'll have a look at it.

**Comment by mstv** [comment]
Process Explorer does not show all spawned Git.exe processes. GE's git command log works better (as long as the PC is responsive): During my current build Git was called every 3.1 seconds with a runtime of about 250-400 ms.

**Comment by RussKie** [comment]
> build Git was called every 3.1 seconds with a runtime of about 250-400 ms.

Err, what? Are you able to trace what spins them?

**Comment by NN---** [comment]
>Is this issue affecting only this repo?
I don't know, possibly there is something in the repo.
I can collect performance data with dotTrace or smth if it is better.
What information can help ?

**Comment by gerhardol** [comment]
>What information can help ?

Git Command Log

>> build Git was called every 3.1 seconds with a runtime of about 250-400 ms.

>Err, what? Are you able to trace what spins them?

This is normal, see https://github.com/gitextensions/gitextensions/issues/6521#issuecomment-489331178
git-status may use up to 1/3 CPU of one core when building, one git process at a time.

**Comment by RussKie** [comment]
> I can collect performance data with dotTrace or smth if it is better.
> What information can help ?

I don't know at this stage. 
Something that could get at least one step closer to understanding what's spinning all these git processes.

Is this happening when you are 
- opening a repo (and not doing anything else)?
- browsing commits?
- viewing diffs?
- committing?
- pulling/pushing?

**Comment by NN---** [comment]
Seems that opening the repository is good enough to reproduce.

**Comment by NN---** [comment]
I think when I open my repository first time after clone it happens.
I am going to double check it and if I can reproduce , collect dotTrace or WPR, whatever helps you.
Is there any option for logs file output maybe ?

**Comment by gerhardol** [comment]
Press F12 to to get the Git Command Log to get the output
However, if you have Tortoise installed too with shell integration, have you checked if the problem occurs if GE is not running? (The tgit indexer used to be very heavy after a clone.)

**Comment by NN---** [comment]
I reproduced issue without tortoisegit.

**Comment by NN---** [comment]
I have another reproduction.
Opened repository, opened branches list, right click on branch.
GitExtensions started spawning git.exe and git-lfs.exe processes without stopping.
Only after killing git.exe processes I could checkout the branch.

**Comment by pmiossec** [comment]
>GitExtensions started spawning git.exe and git-lfs.exe processes without stopping.
>Only after killing git.exe processes I could checkout the branch.

Can't it be a `git gc` ?

I think it's worth checking before killing the processes.

**Comment by NN---** [comment]
Sent file with profiling to the mail.
Pass: gitext

**Comment by VultureX** [comment]
I have this as well. Once I open my repository a certain command is running all the time (see screenshot). Multiple instances of it are being spawned and it takes 3 seconds to complete so CPU usage is high the whole time. Once I open a commit window this command stops being executed and CPU usage returns to normal. Once the commit window is closed it starts using CPU again

![image](https://user-images.githubusercontent.com/5485033/59965551-048ee600-9510-11e9-950b-9589e851873c.png)


This is with
- Git Extensions 3.1.1.6049
- Build 2f872105b4eb42421d9fdafd68e90a9e27d20d96
- Git 2.20.1.windows.1
- Microsoft Windows NT 10.0.17763.0
- .NET Framework 4.7.3362.0
- DPI 144dpi (150% scaling)

**Comment by VultureX** [comment]
I've reverted back to 
- Git Extensions 3.0.2.5232
- Build 23b6f51905006ccdde8701591df706284b4155dc
- Git 2.20.1.windows.1
- Microsoft Windows NT 10.0.17763.0
- .NET Framework 4.7.3362.0
- DPI 144dpi (150% scaling)

This version runs also runs this command periodically, but way less often so there is no excessive CPU usage continuously.

**Comment by pmiossec** [comment]
I don't know if #5983 is the cause of the regression.

Because the status is not written in the index (to prevent index file not writable because used by other commands) when doing 'git status', it is doing all the times the same costly processings taking at each call a lot of Cpu for big repositories. 

I didn't find a clear answer to this problematic but [this link](https://public-inbox.org/git/20170921043214.pyhdsrpy4omy54rm@sigill.intra.peff.net/) seems to give information that try to confirm it....

If you could build your own version of git extensions, I will be pleased if you could test by removing the '--no-optional-lock' argument and give us the result.

**Comment by gerhardol** [comment]
> I don't know if #5983 is the cause of the regression.

Should not be, --no-optional-locks was added to git-status earlier in 2.51. This was adding the flag in a few other background commands (that probably do not set any locks anyway)

GE starts git-status when the file system watcher detects changes. There should only be one git-status running at a time (with exceptions). The time to next command is the largest of 3s or 3 times the time to run the command. There is no queue, just the time when next command is allowed to run.
The start of the log would have been interesting here, how was this situation triggered?

There is only a minor change in GitStatusMonitor for 3.1, for async handling. This was part of improved handling for submodule status that could give performance issues in 3.0.

The situation where a new command can be started also if another is running is for instance when you force refresh, so then a new command is allowed. The current git-status cannot be killed (GE cannot do that) but a new git-status can be started. These extra commands should not reschedule extra git-status, maybe that is what is happening. Will investigate.

**Comment by gerhardol** [comment]
I submitted a PR. Please try the portable build.

**Comment by RussKie** [comment]
@VultureX I've seen a similar behaviour at one of the previous work places where AV and other corporate bloatware made laptop unusable and very slow...

**Comment by RealityVoid** [comment]
I still see this on V3.2.1, but there it seems that the particular PR referenced here was already merged there, and I think the issue was nor really truly fixed. 

Same exact behavior, same huge number of spawned processed, high CPU load because of them, the F12 git command log looks erely similar.  I initially though there was some weird interaction with Windows Defender but disabling it it seems I get the same behavior. 

Interestingly, I have gitex installed on 2 machines and it only happens on one of them. The one where this reproduces the repo is huuuge and contains a lot of LFS objects. I see _a lot_ of git lfs processed being launched.

**Comment by gerhardol** [comment]
What was addressed in 3.2 was a situation where git-status could start multiple times. I could not reproduce the problem but could see a way it could occur if git-status required more than 5s and a build or similar updated the repo. The workaround was to run git-status in the foreground once, for instance open commit dialog.

If git-lfs runs, this is something else. Open a new issue with git command log and information about the repo.

**Comment by NN---** [comment]
It started happening everytime whenever I open a repository in Git Extensions.
I'll try recording with Process Monitor which git commands Git Extensions runs.

**Comment by RussKie** [comment]
is this issue specific to a repo?

**Comment by NN---** [comment]
Yes. It happens in repo with 
lfs files.

**Comment by gerhardol** [comment]
If it is lfs related, open a new issue (with sufficient info like git command log).
Regardless of what the problem was in the original post, there was not enough information and another change has been connected to this issue.

**Comment by NN---** [comment]
This what I get, a loop of git.exe , sh.exe and git-lfs.exe
![image](https://user-images.githubusercontent.com/580536/69331683-e44d8980-0c5d-11ea-9bed-0ed130358544.png)

**Comment by gerhardol** [comment]
Git Command Log is still missing

**Comment by NN---** [comment]
This one ?

```
"C:\Program Files\Git\bin\git.exe" -c log.showSignature=false log -z --pretty=format:"%H%T%P%n%at%n%ct%n%e%n%aN%n%aE%n%cN%n%cE%n%s%n%n%b1DEA7CC4-FB39-450A-8DDF-762FCEA28B05" --all --boundary --not --glob=notes --not --max-count=100000 --```

![image](https://user-images.githubusercontent.com/580536/69335287-452c9000-0c65-11ea-9e6e-35c1e0f5435b.png)

**Comment by pmiossec** [comment]
Also, it could help if you could find by which git command the git lfs
commands are run (git status?).

Process explorer could help,i think

**Comment by gerhardol** [comment]
F12, save the log
See comments in the beginning of this issue

**Comment by NN---** [comment]
This is the log and the last command is keeping repeating itself .

```
2019-11-21T14:19:21.0683963+02:00	119	36732	UI	1	git	config --get ą	
2019-11-21T14:19:21.1985246+02:00	129	20348	UI	1	git		
2019-11-21T14:19:21.7385552+02:00	127	51756	UI	0	git	--version	
2019-11-21T14:19:21.8682644+02:00	204	57140		128	git	--no-optional-locks for-each-ref --sort=-committerdate refs/heads/ --format="%(objectname) %(refname)"	
2019-11-21T14:19:21.8682644+02:00	202	58948		128	git	--no-optional-locks show-ref --dereference	
2019-11-21T14:19:21.8682644+02:00	211	57380		128	git	show-ref --tags	
2019-11-21T14:19:21.9334597+02:00	194	59128	UI	128	git	rev-parse --git-common-dir	
2019-11-21T14:19:22.0731840+02:00	255	58928		128	git	remote -v	
2019-11-21T14:19:22.0751801+02:00	253	58920		128	git	for-each-ref --format="%(push:track)::%(upstream:track)::%(refname:short)" refs/heads/	
2019-11-21T14:19:22.3295193+02:00	239	55620		128	git	show-ref --dereference	
2019-11-21T14:19:22.3614340+02:00	242	48684		128	git	symbolic-ref HEAD	
2019-11-21T14:19:22.5697010+02:00	188	61812		128	git	remote	
2019-11-21T14:19:43.8385630+02:00	268	61760		0	git	show-ref --tags	C:\Test
2019-11-21T14:19:43.8385630+02:00	304	60152		0	git	--no-optional-locks show-ref --dereference	C:\Test
2019-11-21T14:19:43.8385630+02:00	268	46536		0	git	--no-optional-locks for-each-ref --sort=-committerdate refs/heads/ --format="%(objectname) %(refname)"	C:\Test
2019-11-21T14:19:43.8385630+02:00	264	51324	UI	0	git	rev-parse HEAD	C:\Test
2019-11-21T14:19:44.1084486+02:00	173	4680	UI	0	git	rev-parse --git-common-dir	C:\Test
2019-11-21T14:19:44.1134932+02:00	187	52920		0	git	for-each-ref --format="%(push:track)::%(upstream:track)::%(refname:short)" refs/heads/	C:\Test
2019-11-21T14:19:44.1487131+02:00	162	46496		0	git	remote -v	C:\Test
2019-11-21T14:19:44.2882751+02:00	142	58576	UI	0	git	rev-parse HEAD	C:\Test
2019-11-21T14:19:44.3184122+02:00	166	56192		0	git	show-ref --dereference	C:\Test
2019-11-21T14:19:44.4283880+02:00	195	62144	UI	0	git	show-ref --dereference	C:\Test
2019-11-21T14:19:44.4883117+02:00	178	60704		0	git	remote	C:\Test
2019-11-21T14:19:44.6484063+02:00	1054	40956			git	-c log.showSignature=false log -z --pretty=format:"%H%T%P%n%at%n%ct%n%e%n%aN%n%aE%n%cN%n%cE%n%s%n%n%b1DEA7CC4-FB39-450A-8DDF-762FCEA28B05" --all --boundary --not --glob=notes --not --max-count=100000 --	C:\Test
2019-11-21T14:19:44.6783340+02:00	318	18300		0	git	for-each-ref --sort=-committerdate refs/heads/ --format="%(objectname) %(refname)"	C:\Test
2019-11-21T14:19:44.6882826+02:00	290	56300		0	git	ls-files -z --unmerged	C:\Test
2019-11-21T14:19:45.8582956+02:00	1277	56996	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:19:45.9186055+02:00	295	62268		0	git	branch --contains 2e52dbaef365477677f6291f1c754b17a76a641b	C:\Test
2019-11-21T14:19:45.9186055+02:00	304	49388		0	git	for-each-ref --sort=-committerdate --sort=-taggerdate --format="%(refname)" refs/	C:\Test
2019-11-21T14:19:46.2738915+02:00	361	58520		0	git	tag --contains 2e52dbaef365477677f6291f1c754b17a76a641b	C:\Test
2019-11-21T14:19:46.6386840+02:00	332	57776		0	git	describe --tags --first-parent --abbrev=40 2e52dbaef365477677f6291f1c754b17a76a641b	C:\Test
2019-11-21T14:19:47.3386232+02:00	957	61980	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:19:48.5487412+02:00	915	60784	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:19:49.6587056+02:00	914	63868	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:19:50.7783934+02:00	1044	61720	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:19:51.9989063+02:00	985	61308	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:19:53.2185545+02:00	1007	54820	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:19:54.4389899+02:00	963	48040	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:19:55.6585746+02:00	893	63988	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:19:56.7553121+02:00	1250	53324	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:19:58.1986833+02:00	1184	61224	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:19:59.6383669+02:00	1054	61764	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:00.9782963+02:00	1039	62388	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:02.2988581+02:00	899	26260	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:03.4392310+02:00	949	60424	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:04.6683992+02:00	930	17900	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:05.8733679+02:00	965	62760	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:07.1190659+02:00	982	64340	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:08.3388255+02:00	917	52260	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:09.4386698+02:00	969	64260	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:10.6591680+02:00	936	43816	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:11.8836306+02:00	924	61740	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:12.9886978+02:00	879	62976	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:14.0887477+02:00	922	57776	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:15.3089697+02:00	979	62824	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:16.6404522+02:00	1087	61268	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:17.9685834+02:00	1015	56480	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:19.2285265+02:00	969	34492	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
2019-11-21T14:20:20.4387583+02:00	956	51756	UI	0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	C:\Test
```

**Comment by gerhardol** [comment]
There are no 'overlapping' runs of git-status, each invocation is finished when next run starts. The time until next runs start is consistent with the 'interactive' delay time of 200ms plus the periodic timer that runs ever 100 ms. What was changed was a risk that several git-status ran in parallel.

git-status should start 'soon' after a opening a repo or a refresh, which is implemented as a delay of 200ms and the periodic timer runs every 100 ms. After that git-status runs if there are changes in the file system, running at most every 3rd second (or 3 times git-status if larger than 3s).

So the behavior is much like constantly pressing Refresh (or F5). It is possible for plugins to do this too. At least BackgroundFetch and Gerrit. What plugins do you use @NN--- ? Anything else?

**Comment by NN---** [comment]
I deleted Plugins folder but this behavior still continues.

**Comment by NN---** [comment]
This is the call stack:
```
   at GitCommands.Logging.CommandLog.LogProcessStart(String fileName, String arguments, String workDir)
   at GitCommands.Executable.ProcessWrapper..ctor(String fileName, String arguments, String workDir, Boolean createWindow, Boolean redirectInput, Boolean redirectOutput, Encoding outputEncoding)
   at GitCommands.Executable.Start(ArgumentString arguments, Boolean createWindow, Boolean redirectInput, Boolean redirectOutput, Encoding outputEncoding)
   at GitCommands.ExecutableExtensions.<GetOutputAsync>d__3.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.Start[TStateMachine](TStateMachine& stateMachine)
   at GitCommands.ExecutableExtensions.GetOutputAsync(IExecutable executable, ArgumentString arguments, Byte[] input, Encoding outputEncoding, CommandCache cache, Boolean stripAnsiEscapeCodes)
   at GitUI.CommandsDialogs.BrowseDialog.GitStatusMonitor.<>c__DisplayClass39_0.<<Update>b__0>d.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine](TStateMachine& stateMachine)
   at GitUI.CommandsDialogs.BrowseDialog.GitStatusMonitor.<>c__DisplayClass39_0.<Update>b__0()
   at Microsoft.VisualStudio.Threading.JoinableTaskFactory.ExecuteJob[T](Func`1 asyncMethod, JoinableTask job)
   at Microsoft.VisualStudio.Threading.JoinableTaskFactory.RunAsync(Func`1 asyncMethod, Boolean synchronouslyBlocking, JoinableTaskCreationOptions creationOptions, Delegate entrypointOverride)
   at GitUI.CommandsDialogs.BrowseDialog.GitStatusMonitor.Update()
   at System.Windows.Forms.Timer.OnTick(EventArgs e)
   at System.Windows.Forms.Timer.TimerNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
   at System.Windows.Forms.UnsafeNativeMethods.DispatchMessageW(MSG& msg)
   at System.Windows.Forms.UnsafeNativeMethods.DispatchMessageW(MSG& msg)
   at System.Windows.Forms.Application.ComponentManager.System.Windows.Forms.UnsafeNativeMethods.IMsoComponentManager.FPushMessageLoop(IntPtr dwComponentID, Int32 reason, Int32 pvLoopData)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(Int32 reason, ApplicationContext context)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoop(Int32 reason, ApplicationContext context)
   at GitExtensions.Program.RunApplication()
   at GitExtensions.Program.Main()
```

**Comment by gerhardol** [comment]
> This is the call stack:

GitStatusMonitor is the only part that should use git-status --no-optional-locks
The handling seem as it is triggered constantly why GitStatusMonitor 'refreshes' constantly.
There should be other parts of the GUI refreshing too.

**Comment by Newbies42** [comment]
Hello all,
I come back about this subjet because I have the same problem too. I did many search on Internet and I never find any solution.
Here is the list of different topics than I think which could be linked to this issue :
https://github.com/Microsoft/vscode/issues/71085
https://github.com/git-for-windows/git/issues/3318
https://github.com/gitextensions/gitextensions/issues/230

This subject was closed but what is the solution ? 

Thank you for your reply.
